web/packages/teleport/src/services/api/parseError.ts

```diff
@@ -31,11 +31,21 @@ export default function parseError(json) {
 
 export class ApiError extends Error {
   response: Response;
+  messages: string[];
 
-  constructor(message: string, response: Response, opts?: ErrorOptions) {
+  constructor(
+    message: string,
+    response: Response,
+    opts?: ErrorOptions,
+    // messages contains a list of other related
+    // errors besides the main error set for field
+    // "message".
+    messages?: string[]
```

Comment:
> Sometimes the backend can respond with a list of other related error messages but in another field called `messages`

Could you share an example use case for this? I want to understand what exactly `messages` will be used for as it might be a good idea to establish a more concrete error hierarchy.

At the moment, I see that we have `ApiError` which we use in two situations:

* When JSON parsing fails to return an error with an arbitrary message.

https://github.com/gravitational/teleport/blob/30e0e2f1baf83c8ae3f08999be7aa6b27ecbdcc2/web/packages/teleport/src/services/api/api.ts#L115-L118

* On a regular error response to pluck out the error out of the JSON.

https://github.com/gravitational/teleport/blob/30e0e2f1baf83c8ae3f08999be7aa6b27ecbdcc2/web/packages/teleport/src/services/api/api.ts#L131

I assume that `messages` will be for some other third kind of error which might be returned from some endpoints. In that case, it could be beneficial to have something like `ApiError`, `ResponseError`, `ResponseWithMessagesError` where each subsequent class extends the previous one and adds some extra stuff. This would have a couple of benefits:

* Callsites which need to handle errors can match on the error class (`err instanceof SomeClass`) and have specific expectations with regards to what properties are available on a specific error class.
  * After we add `messages`, if there's a callsite that matches on `ResponseError`, it's clear that it doesn't care about extra properties like `messages` and just deals with regular errors returned from pretty much all API endpoints.
* The logic behind `parseError` could be encapsulated within the constructor of `ResponseError` instead of being forced to pluck out some stuff out of JSON manually before creating a new instance of the error.

---

I started writing this digression about error hierarchy, but I want to come back to the actual use case behind `messages`. If it's not going to be commonly used and the field is added only for the purpose of a single endpoint, perhaps it's better to change how the success responses from that endpoint look like so that `messages` can be included in a success response instead? If I knew what the actual use case for `messages` is, I could offer a more concrete advice here.

What I'm saying is that if we need to handle a single specific scenario of a single endpoint, it might be better to handle that through the success response. Otherwise every place in the app would have to consider the fact that `ApiError` might include `messages`, despite the fact that there's only a single callsite that has to ever deal with that.

Reply:
> If I knew what the actual use case for messages is, I could offer a more concrete advice here.

[Here](https://github.com/gravitational/teleport.e/pull/3388/files#diff-bef75e060c3b2a2f83f2b4b5fa6b877e80ac2c1d329a4b16a6b4f3753a6b9f2aR237) is how i used it. users are allowed to define a list of regex exp using `react creatable` (the regex are used to filter okta groups or apps returned by the okta api). 

The main error we display to user (the cause of error) is from `error.message`, and then internally we check all error messages to match the filter that caused an error and highlight the offending filter inside the `react creatable input box`:

<img width="593" alt="image" src="https://github.com/gravitational/teleport/assets/43280172/2269439f-03c8-46f4-83e4-f5079d8cc47d">

example error response:
![image](https://github.com/gravitational/teleport/assets/43280172/bfca6963-1e65-4c1f-9244-c47f36a944a6)

>  it could be beneficial to have something like ApiError, ResponseError, ResponseWithMessagesError 

After i explained the use case do you think it's still applicable? i thought being a part of `ApiError` makes sense as all endpoints could potentially [return it](https://github.com/gravitational/trace/blob/940ad545a1410e236ba5cd149f3e245a26feb387/trace.go#L243) ([this](https://github.com/gravitational/teleport.e/blob/ade1f1b9558be87729f6a22466f1f2973c1d6ee9/lib/okta/pluginconfighelper.go#L210) is how it gets formed)

Reply:
Oh, I didn't know that trace stores all messages from `trace.Wrap` on the error object like this. If so then I think it's totally okay to have this on ApiError.